packages/server/test/preHook.test.ts

```diff
@@ -0,0 +1,167 @@
+import { routerToServerAndClient } from './_testHelpers';
+import * as trpc from '../src';
+import { httpError } from '../src';
+
+test('is called if def first', async () => {
+  const preHook = jest.fn();
+  const { client, close } = routerToServerAndClient(
+    trpc
+      .router()
+      .preHook(preHook)
+      .query('foo1', {
+        resolve() {
+          return 'bar1';
+        },
+      })
+      .query('foo2', {
+        resolve() {
+          return 'bar2';
+        },
+      }),
+  );
+
+  expect(await client.query('foo1')).toBe('bar1');
+  expect(await client.query('foo2')).toBe('bar2');
+  expect(preHook).toHaveBeenCalledTimes(2);
+  close();
+});
+
+test('is not called if def last', async () => {
+  const preHook = jest.fn();
+  const { client, close } = routerToServerAndClient(
+    trpc
+      .router()
+      .query('foo', {
+        resolve() {
+          return 'bar';
+        },
+      })
+      .preHook(preHook),
+  );
+
+  expect(await client.query('foo')).toBe('bar');
+  expect(preHook).toHaveBeenCalledTimes(0);
+  close();
+});
+
+test('allows you to throw an error (e.g. auth)', async () => {
+  type Context = {
+    user?: {
+      id: number;
+      name: string;
+      isAdmin: boolean;
+    };
+  };
+  const resolverMock = jest.fn();
+
+  const headers: Record<string, string | undefined> = {};
```

Comment:
doesnt @types/node come with some Headers typing you can import?

Reply:
i don't care enough about test code to care about changing this :)

there's some reason why i don't use that elsewhere, but i don't remember why

Reply:
ye I am the same with tests. Doesnt need to be as readable. I just figured while you writing its easier to import heders than to type the thing manually but its already done, no need to change it now